Add initial CWL JSON schema definition

fmigneault opened the pull request:
Add a "minimalistic" (as not fully feature complete, but relatively high coverage of most common definitions) CWL schema using JSON schema representation. 

The main schema definition of interest is `DeployCWL` which represents one of the content combintion that should be POST'd to `/processes` to deploy using CWL in *OGC API - Processes - Part 2: Deploy, Replace, Update (DRU)*.

We are also working on porting the [`EchoProcess`](https://github.com/opengeospatial/ogcapi-processes/blob/24e0e17b4d6773aa856500f395109f29e92f00c7/core/examples/json/ProcessDescription.json) to its CWL equivalent (https://github.com/crim-ca/weaver/blob/implement-example-process/weaver/processes/builtin/echo_process.cwl). This could be added as example as well.

Relates to https://github.com/opengeospatial/ogcapi-processes/issues/319

mr-c commented on openapi/schemas/processes-dru/cwl.yaml:
```diff
@@ -0,0 +1,1650 @@
+    description: | 
+      Hint indicating that the Application Package corresponds to an
+      OGC API - Processes provider that should be remotely executed and monitored
+      by this instance. (note: can only be an 'hint' as it is unofficial CWL specification).
```
If the process is not otherwise executable without making use of the `OGCAPIRequirement` then it should be under `requirements` not `hints`. Non-standard extensions are allowed in both fields

fmigneault replied:
I agree. 
I think it is implementation dependent whether this would be required or optional on the target platform where the process is being deployed. However, I would like to make this definition proposal somewhat established such that different implementators could be guided toward a somewhat interoperable reference.

mr-c commented on openapi/schemas/processes-dru/cwl.yaml:
```diff
@@ -0,0 +1,1650 @@
+    - $ref: '#/$definitions/CWLRequirementsList'
+    title: CWLRequirements
+    description: Explicit requirement to execute the application package.
+  CWLRequirementsMap:
+    anyOf:
+    - $ref: '#/$definitions/DockerRequirementMap'
+    - $ref: '#/$definitions/DockerGpuRequirementMap'
+    - $ref: '#/$definitions/InitialWorkDirRequirementMap'
+    - $ref: '#/$definitions/InlineJavascriptRequirementMap'
+    - $ref: '#/$definitions/InplaceUpdateRequirementMap'
+    - $ref: '#/$definitions/LoadListingRequirementMap'
+    - $ref: '#/$definitions/NetworkAccessRequirementMap'
+    - $ref: '#/$definitions/ResourceRequirementMap'
+    - $ref: '#/$definitions/ScatterFeatureRequirementMap'
+    - $ref: '#/$definitions/ToolTimeLimitRequirementMap'
+    - $ref: '#/$definitions/WorkReuseRequirementMap'
+    - $ref: '#/$definitions/UnknownRequirementMap'
+    title: CWLRequirementsMap
```
```suggestion
  CWLRequirementsMap:
    anyOf:
    - $ref: '#/$definitions/DockerRequirementMap'
    - $ref: '#/$definitions/DockerGpuRequirementMap'
    - $ref: '#/$definitions/InitialWorkDirRequirementMap'
    - $ref: '#/$definitions/InlineJavascriptRequirementMap'
    - $ref: '#/$definitions/InplaceUpdateRequirementMap'
    - $ref: '#/$definitions/LoadListingRequirementMap'
    - $ref: '#/$definitions/NetworkAccessRequirementMap'
    - $ref: '#/$definitions/ResourceRequirementMap'
    - $ref: '#/$definitions/ScatterFeatureRequirementMap'
    - $ref: '#/$definitions/ToolTimeLimitRequirementMap'
    - $ref: '#/$definitions/WorkReuseRequirementMap'
    - $ref: '#/$definitions/CUDARequirementMap'
    - $ref: '#/$definitions/UnknownRequirementMap'
    title: CWLRequirementsMap
```

fmigneault replied:
Applied in e5547f004fdd90ca015b11b10a2700a068b39fce

mr-c commented on openapi/schemas/processes-dru/cwl.yaml:
```diff
@@ -0,0 +1,1650 @@
+  CWLRequirementsMap:
+    anyOf:
+    - $ref: '#/$definitions/DockerRequirementMap'
+    - $ref: '#/$definitions/DockerGpuRequirementMap'
```
Oh, are people still using `DockerGpuRequirement` instead of `CUDARequirement`?

fmigneault replied:
I thought that I had removed it. Let me double check because CRIM also uses and support `CUDARequirement`.

fmigneault replied:
Done. Forgot to remove that reference but the definition was already removed.

mr-c reviewed (commented):
Thank you @fmigneault ; this looks like a lot of work!

A bonus would be to try to load each of the CWL v1.2 conformance tests (except those with `should_fail: true`) with this schema to ensure that there are no obvious things missing. Either as a CI process or a one-time test.

https://github.com/common-workflow-language/cwl-v1.2/blob/1.2.1_proposed/conformance_tests.yaml

fmigneault commented:
@mr-c 

Coverage so far: 
```
================ 188 failed, 185 passed, 373 warnings in 28.33s ================
```


```python
import logging
import os
import posixpath
from functools import cache
from typing import Dict, List, Optional, TypeAlias, TypedDict, TypeGuard, Union
from typing_extensions import NotRequired
from urllib.parse import urljoin, urlparse

import jsonschema
import requests
import pytest
import yaml
from jsonschema.exceptions import ValidationError
from yaml.scanner import ScannerError

CONFORMANCE_TESTS_FILE = "https://raw.githubusercontent.com/common-workflow-language/cwl-v1.2/1.2.1_proposed/conformance_tests.yaml"
CWL_JSON_SCHEMA_FILE = "https://raw.githubusercontent.com/crim-ca/ogcapi-processes/cwl-schema/openapi/schemas/processes-dru/cwl.yaml#$definitions/CWL"

LOGGER = logging.getLogger(__name__)

# pylint: disable=C0103,invalid-name
Number = Union[int, float]
ValueType = Union[str, Number, bool]
AnyValueType = Optional[ValueType]
# add more levels of explicit definitions than necessary to simulate JSON recursive structure better than 'Any'
# amount of repeated equivalent definition makes typing analysis 'work well enough' for most use cases
_JSON: TypeAlias = "JSON"
_JsonObjectItemAlias: TypeAlias = "_JsonObjectItem"
_JsonListItemAlias: TypeAlias = "_JsonListItem"
_JsonObjectItem = Dict[str, Union[AnyValueType, _JSON, _JsonObjectItemAlias, _JsonListItemAlias]]
_JsonListItem = List[Union[AnyValueType, _JSON, _JsonObjectItem, _JsonListItemAlias]]
_JsonItem = Union[AnyValueType, _JSON, _JsonObjectItem, _JsonListItem]
JSON = Union[Dict[str, Union[_JSON, _JsonItem]], List[Union[_JSON, _JsonItem]], AnyValueType]

ConformanceTestDef = TypedDict(
    "ConformanceTestDef",
    {
        "id": str,
        "doc": str,
        "tags": List[str],
        "tool": str,
        "job": NotRequired[str],  # not used, for running the actual CWL
        "output": JSON,  # not used, output of CWL execution
        "should_fail": NotRequired[bool],  # indicates failure as "execute failing", but still a valid CWL
    }
)


@cache
def is_remote_file(file_location: str) -> TypeGuard[str]:
    """
    Parses to file location to figure out if it is remotely available or a local path.
    """
    cwl_file_path_or_url = file_location.replace("file://", "")
    scheme = urlparse(cwl_file_path_or_url).scheme
    return scheme != "" and not posixpath.ismount(f"{scheme}:")  # windows partition


@cache
def load_file(file_path: str, text: bool = False) -> Union[JSON, str]:
    """
    Load :term:`JSON` or :term:`YAML` file contents from local path or remote URL.

    If URL, get the content and validate it by loading, otherwise load file directly.

    :param file_path: Local path or URL endpoint where file to load is located.
    :param text: load contents as plain text rather than parsing it from :term:`JSON`/:term:`YAML`.
    :returns: loaded contents either parsed and converted to Python objects or as plain text.
    :raises ValueError: if YAML or JSON cannot be parsed or loaded from location.
    """
    try:
        if is_remote_file(file_path):
            headers = {"Accept": "text/plain"}
            resp = requests.get(file_path, headers=headers)
            if resp.status_code != 200:
                raise ValueError("Loading error: [%s]", file_path)
            return resp.content if text else yaml.safe_load(resp.content)
        with open(file_path, mode="r", encoding="utf-8") as f:
            return f.read() if text else yaml.safe_load(f)
    except OSError as exc:
        LOGGER.debug("Loading error: %s", exc, exc_info=exc)
        raise
    except ScannerError as exc:  # pragma: no cover
        LOGGER.debug("Parsing error: %s", exc, exc_info=exc)
        raise ValueError("Failed parsing file content as JSON or YAML.")


def load_conformance_tests(test_file: str) -> List[ConformanceTestDef]:
    conformance_tests = load_file(test_file)
    assert isinstance(conformance_tests, list)
    assert all(isinstance(conf_test, dict) for conf_test in conformance_tests)
    test_base = os.path.dirname(test_file)
    all_conf_test = []
    for conf_test in conformance_tests:
        if "$import" in conf_test:
            conf_path = urljoin(f"{test_base}/", conf_test["$import"])
            conf_test = load_conformance_tests(conf_path)
            all_conf_test.extend(conf_test)
            continue
        for ref in ["job", "tool"]:
            if ref not in conf_test:
                continue
            if not urlparse(conf_test[ref]).scheme:
                conf_test[ref] = urljoin(f"{test_base}/", conf_test[ref])
        all_conf_test.append(conf_test)
    return all_conf_test


@pytest.mark.parametrize(
    ["schema_file", "conformance_test"],
    [
        (CWL_JSON_SCHEMA_FILE, conf_test)
        for conf_test in load_conformance_tests(CONFORMANCE_TESTS_FILE)
    ]
)
def test_conformance(schema_file: str, conformance_test: ConformanceTestDef) -> None:
    instance_file = conformance_test["tool"]
    instance = load_file(instance_file)
    schema_path = []
    schema_ref = ""
    if "#" in schema_file:
        schema_file, schema_ref = schema_file.split("#", 1)
        schema_path = [ref for ref in schema_ref.split("/") if ref]
        schema_ref = f"#{schema_ref}"
    schema_base = schema = load_file(schema_file)
    if schema_path:
        for part in schema_path:
            schema = schema[part]
    
    # ensure local schema can find relative $ref, since the provided reference can be a sub-schema (with "#/...")
    scheme_uri = f"file://{schema_file}" if schema_file.startswith("/") else schema_file
    validator = jsonschema.validators.validator_for(schema_base)
    validator.resolver = jsonschema.RefResolver(base_uri=scheme_uri, referrer=schema_base)
    validator(schema_base).validate(instance, schema)  # raises if invalid

```

Should this be made part of https://github.com/common-workflow-language/cwl-v1.2 directly?
Only items that would be defined here would be:
-  `DeployCWL`,
- `DeployOGCAppPackageCWL`
- `CWLExecutionUnit` 
- some form of extended `requirements`/`hints` for `OGCAPIRequirement`/`WPS1Requirement`/`BuiltinRequirement` (and their map variantion)

mr-c commented:
> @mr-c
> 
> Coverage so far:
> 
> ```
> ================ 188 failed, 185 passed, 373 warnings in 28.33s ================
> ```

You're so fast, thanks!
Are you working through the failures, or should I take a stab?



> Should this be made part of https://github.com/common-workflow-language/cwl-v1.2 directly? Only items that would be defined here would be:
> 
>     * `DeployCWL`,
> 
>     * `DeployOGCAppPackageCWL`
> 
>     * `CWLExecutionUnit`
> 
>     * some form of extended `requirements`/`hints` for `OGCAPIRequirement`/`WPS1Requirement`/`BuiltinRequirement` (and their map variantion)
> 
> 
> And those definitions would refer to the official CWL JSON schema in https://github.com/common-workflow-language/cwl-v1.2.

Yes, we would be happy to host that once the validation passes. I assume it is all hand-written? Maybe we can get an intern or junior dev to build an automated schema-salad to OpenAPI converter based upon your work; to keep it updated with future revisions of the CWL standards.

fmigneault commented:
> Are you working through the failures, or should I take a stab?

I'm about to start checking them. 

> I assume it is all hand-written? 

Its generated from a subset of definitions from https://github.com/crim-ca/weaver/blob/master/weaver/wps_restapi/swagger_definitions.py + a few manual edits to patch some formatting. Many of the documented descriptions are copy-pasted from CWL docs with minor formatting edits. They could reuse a reference document.

fmigneault commented:
Update as of a67e7ec
`=========== 88 failed, 281 passed, 4 xfailed, 371 warnings in 57.31s ===========`

I've not gone through all the remaining failing cases, but I saw a few that did not look really complicated. 
Probably just missing a field here and there. 

For the rest, items not yet supported are: 
- `SchemaDefRequirement`
- `SubWorkflowRequirement`
- `Records` inputs/outputs format

mr-c commented:
> Update as of [a67e7ec](https://github.com/opengeospatial/ogcapi-processes/commit/a67e7ec7c150bdcb238edbf3ece68ef40a4a082b) `=========== 88 failed, 281 passed, 4 xfailed, 371 warnings in 57.31s ===========`

Great progress, I'm impressed! I don't think your schema is "minimalistic" anymore, and that is a good thing!

> I've not gone through all the remaining failing cases, but I saw a few that did not look really complicated. Probably just missing a field here and there.

👍

> For the rest, items not yet supported are:
> 
>     * `SchemaDefRequirement` (xfailed cases)
> 
>     * `SubWorkflowRequirement` (any other technicality to validate than just nesting a `Workflow` as a step?)
Correct, it is just a flag that there might be a nested workflow. 
> 
>     * `Records` inputs/outputs format (I've never used those, I will need to investigate further what they imply)

For all three of these: you probably should just accept them and not validate them too deeply.

fmigneault commented on openapi/schemas/processes-dru/cwl.yaml:
```diff
@@ -0,0 +1,1724 @@
+  CWLOutputStdOutDefinition:
+    description: |
+      Indicates that the data pushed to the standard output stream by the command will be redirected to this CWL output.
+      Can be defined for only one output and must not be combined with 'stdout' at the root of the CWL document.
```
Ok. Just to confirm I understood it.
The top-level `stdout` will set the name/location of the file, and the type `stdout` will indicate which output writes to it?

mr-c replied:
That's basically correct, yes. The use of a top-level `stdout` does not require the use of a `type: stdout` in the outputs section. That stdout file could be consumed by a matching `outputBinding: { glob: … }` construct of a `type: File` or `type: Directory` or similar; or it may be unconsumed entirely.

fmigneault commented:
@mr-c 
Not breaking, but the following 2 items are missing a `doc` field (all others have it):
- https://github.com/common-workflow-language/cwl-v1.2/blob/1.2.1_proposed/tests/conditionals/test-index.yaml#L22-L27
- https://github.com/common-workflow-language/cwl-v1.2/blob/1.2.1_proposed/tests/conditionals/test-index.yaml#L206-L211

mr-c commented on openapi/schemas/processes-dru/cwl.yaml:
```diff
@@ -0,0 +1,1788 @@
+    - type: string
+      title: ID
+      description: Generic identifier.
+      pattern: '^[#.]?[A-Za-zA-Za-z0-9]+(?:[-_.][A-Za-zA-Za-z0-9]+)*$'
```
I'm confused by this `A-Za-zA-Za-z` repetition. Is that a search and replace artifact, or does that accomplish something?

fmigneault replied:
No, it is on purpose. For lowercase/uppercase match. 
My original regexes were in code using the case-insensitive flag, but it is not enabled by default in pure JSON schema.

fmigneault commented:
24a5b8c
`==== 38 failed, 318 passed, 18 xfailed, 533 warnings in 1420.67s (0:23:40) =====`
Slowly getting there.

@mr-c 
Some of the remaining failures are related to https://github.com/common-workflow-language/cwl-v1.2/issues/251
Others with `SubworkflowFeatureRequirement` left to implement.

fmigneault commented:
08fb2c4 
`=========== 23 failed, 351 passed, 374 warnings in 68.59s (0:01:08) ============`
All remaining "fails" are related/caused by: 
- https://github.com/common-workflow-language/cwl-v1.2/issues/251
- https://github.com/common-workflow-language/cwl-v1.2/issues/252
- https://github.com/common-workflow-language/cwl-v1.2/issues/253
- `SubworkflowFeatureRequirement`

fmigneault commented:
bc3063e
`================= 5 failed, 369 passed, 374 warnings in 54.60s =================`
Only remaining items are caused by: 
- https://github.com/common-workflow-language/cwl-v1.2/issues/251
- https://github.com/common-workflow-language/cwl-v1.2/issues/252
- https://github.com/common-workflow-language/cwl-v1.2/issues/253

This should now be fully suported CWL if those issues get addressed! 🚀

fmigneault commented:
@mr-c 
Where should I migrate the CWL schema definition (repo and directory) and the python test function?

mr-c commented:
> @mr-c Where should I migrate the CWL schema definition (repo and directory) and the python test function?

Send a PR to https://github.com/common-workflow-language/cwl-v1.2/ ; thanks!

fmigneault commented:
PR is now open: https://github.com/common-workflow-language/cwl-v1.2/pull/256
Until merged, the actual file can be found here: https://github.com/fmigneault/cwl-v1.2/blob/cwl-json-schema/json-schema/cwl.yaml
The local CWL schema for *OGC API - Processes* is updated starting from 6844490 to use the (expected) remote reference.

fmigneault commented:
Closing this in favour of https://github.com/common-workflow-language/cwl-v1.2/pull/256
Official schema: https://github.com/common-workflow-language/cwl-v1.2/blob/1.2.1_proposed/json-schema/cwl.yaml
